File: cogs/show_table.py
# cogs/cmd_show_table.py
import discord
from discord import app_commands
from discord.ext import commands
from utils.data_manager import load_data
from tabulate import tabulate # Importante não esquecer disto

@app_commands.command(name="show_table", description="Mostra a tabela de um torneio (padrão: ativo)")
async def show_table(interaction: discord.Interaction, name: str = None):
    data = load_data()
    active = name or data.get("active")
    if not active:
        await interaction.response.send_message(" Não há torneio ativo nem nome especificado.", ephemeral=True)
        return

    if active not in data["tournaments"]:
        await interaction.response.send_message(" Esse torneio não existe.", ephemeral=True)
        return

    tour = data["tournaments"][active]
    players = tour["players"]

    if not players:
        await interaction.response.send_message(f" O torneio **{active}** ainda não tem jogadores.", ephemeral=True)
        return

    rows = []
    for nome, st in players.items():
        played = st["played"]
        wins = st["wins"]
        draws = st["draws"]
        losses = st["losses"]
        points = st["points"]
        win_pct = (wins / played * 100) if played > 0 else 0.0
        ponderada = ((wins + 0.5 * draws) / played * 100) if played > 0 else 0.0
        rows.append((nome, played, wins, draws, losses, points, win_pct, ponderada))

    rows.sort(key=lambda r: (-r[5], -r[2], -r[6]))

    headers = ["Pos", "Jogador", "J", "V", "E", "D", "Pts", "Win %", "Win % Ponderada"]
    table = [
        [idx, r[0], r[1], r[2], r[3], r[4], r[5], f"{r[6]:.2f}%", f"{r[7]:.2f}%"]
        for idx, r in enumerate(rows, start=1)
    ]

    text = tabulate(table, headers=headers, tablefmt="pretty")
    await interaction.response.send_message(f"**Tabela: {active}**\n```{text}```")

async def setup(bot: commands.Bot):
    bot.tree.add_command(show_table)

File: src/commands/show-table.js
import {
  ApplicationIntegrationType,
  AttachmentBuilder,
  InteractionContextType,
  MessageFlags,
  SlashCommandBuilder,
} from "discord.js";

import { store } from "../data/json-store.js";
import { buildStandings } from "../domain/standings.js";
import { formatTable } from "../utils/table.js";

export const data = new SlashCommandBuilder()
  .setName("show_table")
  .setDescription("Mostra a tabela de um torneio (padrão: ativo)")
  .setIntegrationTypes(ApplicationIntegrationType.GuildInstall)
  .setContexts(InteractionContextType.Guild)
  .addStringOption((option) =>
    option
      .setName("name")
      .setDescription("Nome do torneio")
      .setMinLength(1)
      .setMaxLength(100)
      .setRequired(false),
  );

export async function execute(interaction) {
  const state = await store.load();
  const requestedTournament = interaction.options.getString("name");
  const active = requestedTournament || state.active;

  if (!active) {
    await interaction.reply({
      content: " Não há torneio ativo nem nome especificado.",
      flags: MessageFlags.Ephemeral,
    });
    return;
  }

  if (!Object.hasOwn(state.tournaments, active)) {
    await interaction.reply({
      content: " Esse torneio não existe.",
      flags: MessageFlags.Ephemeral,
    });
    return;
  }

  const players = state.tournaments[active].players;
  if (Object.keys(players).length === 0) {
    await interaction.reply({
      content: ` O torneio **${active}** ainda não tem jogadores.`,
      flags: MessageFlags.Ephemeral,
    });
    return;
  }

  const standings = buildStandings(players);
  const table = formatTable(
    ["Pos", "Jogador", "J", "V", "E", "D", "Pts", "Win %", "Win % Ponderada"],
    standings.map((row) => [
      row.position,
      row.name,
      row.played,
      row.wins,
      row.draws,
      row.losses,
      row.points,
      `${row.winPercentage.toFixed(2)}%`,
      `${row.weightedWinPercentage.toFixed(2)}%`,
    ]),
  );
  const codeBlockSafeTable = table.replaceAll("```", "``\u200b`");
  const message = `**Tabela: ${active}**\n\`\`\`${codeBlockSafeTable}\`\`\``;

  if (message.length <= 2_000) {
    await interaction.reply(message);
    return;
  }

  const attachment = new AttachmentBuilder(
    Buffer.from(`Tabela: ${active}\n\n${table}\n`, "utf8"),
    { name: "tabela.txt" },
  );
  await interaction.reply({
    content: "A tabela completa está no arquivo anexado.",
    files: [attachment],
  });
}
